### PR TITLE
BUG: Align type definition with generated lapack

### PR DESCRIPTION
**Pull request by ghost**
See gh-10451.

Closes gh-10452.

### CODE REVIEW

**Comment by eric-wieser** [comment]
Commit message is wrong, this is the definition.

Also, you need to hit the non-K&R prototype on the following line.

A comment explaining the purpose of the change would be valuable too.

**Comment by xoviat** [comment]
Yes, this is not declaring a type. Silly me.

**Comment by eric-wieser** [comment]
I think we should try and fix all of these bugs at once so we can get a feel for the scope. Perhaps trying to get the linker warnings to be errors in the one travis build that uses this code?

**Comment by xoviat** [comment]
The only reason that MSVC detected this is that it was generating code at link-time. The errors span translation units--is gcc able to detect these errors? According to C and C specifications, I don't think the compiler is required to detect these errors.

**Comment by charris** [comment]
Close & reopen.

File: numpy/linalg/lapack_lite/f2c.h [hunk 1]
```diff
@@ -214,4 +214,108 @@ typedef doublereal E_f;	/* real function with -R not specified */
...
+#endif
+#if defined(KR_headers)
+float sqrtf(),f__cabsf();
+#endif
+#if !(defined(KR_headers))
+double f__cabsf(float,float);
+#endif
+#if defined(KR_headers)
+double f__cabsf();
+#endif
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/f2c.h]: `#if defined(KR_headers)` seems to be doing the wrong thing. We don't actually need to support this use case anyway - it just happens that `f2c.c` is a concatenation of files from netlib, which are otherwise untouched.

Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c.h]: We can't remove this from the header file unless we remove it from the C file as well.

Comment (eric-wieser) [reply numpy/linalg/lapack_lite/f2c.h]: Ah, just looked it up - what you have is the correct way to declare K&R prototypes

File: numpy/linalg/lapack_lite/f2c.h [hunk 2]
```diff
@@ -214,4 +214,108 @@ typedef doublereal E_f;	/* real function with -R not specified */
...
+#if defined(KR_headers)
+double f__cabs();
+float sqrtf();
+double f__cabsf(real,imag)float real,imag;
+#endif
+#if !(defined(KR_headers))
+double f__cabsf(float,float);
+#endif
+#if defined(KR_headers)
+float sqrtf(),f__cabsf();
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/f2c.h]: This definition doesn't match the other 4 `f__cabsf` definitions...

Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c.h]: That comes from here:

https://github.com/numpy/numpy/blob/04c75bdee87559a7a2e078b69af62136a0679b05/numpy/linalg/lapack_lite/f2c.c#L627

Should I remove it from f2c.c?

Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c.h]: This should actually be a float...

**Comment by xoviat** [comment]
Okay, I think this is ready to merge.

**Comment by eric-wieser** [comment]
Why are we not seeing #r_imag in here? Are they declared with the wrong type, but consistently?

File: numpy/linalg/lapack_lite/f2c.h [hunk 3]
```diff
@@ -214,4 +214,108 @@ typedef doublereal E_f;	/* real function with -R not specified */
 #undef unix
 #undef vax
 #endif
+
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/f2c.h]: Should add a comment here explaining that these are messy because they are generated from the C file (and explain how to generate them)

**Comment by xoviat** [comment]
> Why are we not seeing #10452 in here? Are they declared with the wrong type, but consistently?

`makeheaders` appears to have left out some prototypes. I'm going to try again with a newer version.

**Comment by xoviat** [comment]
Not sure how to proceed. The type definition of the generated files is incorrect, but modifying generated files introduced technical debt.

**Comment by eric-wieser** [comment]
Can't you just modify the non-generated ones to use larger types than necessary, or is the generated code self-inconsistent?

Creating a patch file or a parser to fixup the generated code I think would be the right approach

**Comment by xoviat** [comment]
@eric-wieser I think I've resolved all of the issues. In addition, I've created a new `plex3` package on pypi so that it's possible to regenerate the sources using Python 3.

File: numpy/linalg/lapack_lite/clapack_scrub.py [hunk 4]
```diff
@@ -2,11 +2,17 @@
...
 import re
 from plex import Scanner, Str, Lexicon, Opt, Bol, State, AnyChar, TEXT, IGNORE
 from plex.traditional import re as Re
 
+PY2 = sys.version_info < (3, 0)
+
+if PY2:
+    from io import BytesIO
+else:
+    from io import StringIO as BytesIO
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/clapack_scrub.py]: This looks pretty sketchy - perhaps call it `UStringIO` which would at least match the internal name we use for the python 2/3 string type, or perhaps `TextIO` which I think we use in tests somewhere.

File: numpy/linalg/lapack_lite/clapack_scrub.py [hunk 5]
```diff
@@ -221,6 +227,37 @@ def OutOfHeader(line):
         state = state(line)
     return lines.getValue()
 
+def removeSubroutinePrototypes(source):
+    expression = re.compile(
+        '/\* Subroutine \*/^\s*(?:(?:inline|static)\s+){0,2}(?!else|typedef|return)\w+\s+\*?\s*(\w+)\s*\([^0]+\)\s*;?'
+    )
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/clapack_scrub.py]: Maybe use verbose more here to make this a little easier to read?

File: numpy/linalg/lapack_lite/clapack_scrub.py [hunk 6]
```diff
@@ -221,6 +227,37 @@ def OutOfHeader(line):
...
+
+    def InBuiltInFunctions(line):
+        if line.strip() == '':
+            return LookingForBuiltinFunctions
+        else:
+            return InBuiltInFunctions
+
+    state = LookingForBuiltinFunctions
+    for line in BytesIO(source):
+        state = state(line)
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/clapack_scrub.py]: Neat state machine construct.  Inspired from elsewhere in the file, or just the way you normally do this?

File: numpy/linalg/lapack_lite/f2c_config.c [hunk 7]
```diff
@@ -1773,7 +1759,7 @@ doublereal slamch_(char *cmach)
 doublereal slamc3_(real *a, real *b)
 {
     /* System generated locals */
-    volatile real ret_val;
+    real ret_val;
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/f2c_config.c]: Is this change deliberate?

File: numpy/linalg/lapack_lite/make_lite.py [hunk 8]
```diff
@@ -266,13 +277,26 @@ def scrubF2CSource(c_file):
         fo.write(HEADER)
         fo.write(source)
 
+def ensure_executable(name):
+    try:
+        which(name)
+    except:
+        print(name + ' not found')
+        raise SystemExit()
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/make_lite.py]: Combine these two lines into `raise SystemExit(message)`

File: numpy/linalg/lapack_lite/make_lite.py [hunk 9]
```diff
@@ -232,6 +240,8 @@ def getWrappedRoutineNames(wrapped_routines_file):
 types = {'blas', 'lapack', 'd_lapack', 's_lapack', 'z_lapack', 'c_lapack', 'config'}
 
 def dumpRoutineNames(library, output_dir):
+    if not PY2:
+        os.makedirs(output_dir, exist_ok=True)
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/make_lite.py]: Not sure it's worth doing this if we only do it on py3, nor am I sure that creating the output directory is in-scope for what the build script should do.

File: numpy/linalg/lapack_lite/make_lite.py [hunk 10]
```diff
@@ -251,6 +261,7 @@ class F2CError(Exception):
     pass
 
 def runF2C(fortran_filename, output_dir):
+    fortran_filename = fortran_filename.replace('\\', '/')
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/make_lite.py]: Why's this needed, and should it be using os.pathsep?

File: numpy/linalg/lapack_lite/f2c.h [hunk 11]
```diff
@@ -214,4 +214,171 @@ typedef doublereal E_f;	/* real function with -R not specified */
...
+extern void sig_die(char *, int);
+extern integer signal_(integer *, void (*)(int));
+extern integer system_(char *, ftnlen);
+extern double z_abs(doublecomplex *);
+extern void z_cos(doublecomplex *, doublecomplex *);
+extern void z_div(doublecomplex *, doublecomplex *, doublecomplex *);
+extern void z_exp(doublecomplex *, doublecomplex *);
+extern void z_log(doublecomplex *, doublecomplex *);
+extern void z_sin(doublecomplex *, doublecomplex *);
+extern void z_sqrt(doublecomplex *, doublecomplex *);
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/f2c.h]: I'm a little wary of the fact that either:
* You created this manually
* You created this with an automated process that isn't documented in a way that can be repeated in future.

Should we perhaps have the script check the list of headers in this file against the ones it removed from the source, and print out warnings/errors if some are missing?

**Comment by eric-wieser** [comment]
Nice approach of just removing all of the inline header

File: numpy/linalg/lapack_lite/f2c_config.c (hunk 7 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c_config.c]: This was introduced during regeneration. The variable is declared locally and isn't returned before the end of the function, so I personally can't see a reason why this routine wouldn't be subject to optimization.

File: numpy/linalg/lapack_lite/make_lite.py (hunk 9 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/make_lite.py]: `exist_ok` doesn't exist on PY2, and I don't enjoy making up for its shortcomings.

File: numpy/linalg/lapack_lite/make_lite.py (hunk 10 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/make_lite.py]: f2c cannot understand `\\`, even if it's compiled on windows.

File: numpy/linalg/lapack_lite/make_lite.py (hunk 8 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/make_lite.py]: Okay.

File: numpy/linalg/lapack_lite/python_xerbla.c [hunk 12]
```diff
@@ -1,4 +1,6 @@
 #include "Python.h"
+
+#undef c_abs
```
[the remark this answers is not in the document]
Comment (xoviat) [reply numpy/linalg/lapack_lite/python_xerbla.c]: `c_abs` is defined in Python.h. Therefore, calling `c_abs` when linked to both `libpython` and `libf2c` is ambiguous. Updating the headers caused the compiler to detect this ambiguity and throw an error.

File: numpy/linalg/lapack_lite/f2c.h (hunk 11 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c.h]: These headers come directly from `libf2c` (specifically, some random gh repository).

> Should we perhaps have the script check the list of headers in this file against the ones it removed from the source, and print out warnings/errors if some are missing?

The compiler can check this as long as there are no inline prototypes (which is why inline prototypes are a really bad idea). The only script that we need to write is one which will remove the inline prototypes, which is what I have done here.

File: numpy/linalg/lapack_lite/f2c.h (hunk 3 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c.h]: As explained below, these are copypasted from `libf2c`, and the compiler will detect (and does detect) any missing/incorrect definitions (see all of the new warnings that the compiler previously missed). I this case, I don't believe that time should be spent on the aesthetic issue, but rather the compiler warnings (and in another PR).

File: numpy/linalg/lapack_lite/clapack_scrub.py (hunk 6 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/clapack_scrub.py]: Essentially copypasted from `removeHeader` above.

File: numpy/linalg/lapack_lite/clapack_scrub.py (hunk 4 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/clapack_scrub.py]: Okay.

**Comment by xoviat** [comment]
@eric-wieser Is there anything else before we can merge?

File: numpy/linalg/lapack_lite/f2c_config.c (hunk 7 above)
Comment (eric-wieser) [reply numpy/linalg/lapack_lite/f2c_config.c]: I'm a little concerned that the regeneration produced a different result. What's your F2C version and lapack source version?

Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c_config.c]:
```
f2c (Fortran to C Translator) version 20100827.
```

LAPACK version is the one specified in the README (3.2.2 I think).

Comment (eric-wieser) [reply numpy/linalg/lapack_lite/f2c_config.c]: That's the same version as mine...

**Comment by eric-wieser** [comment]
I want to check this out and build on my machine, and verify the generated files are unchanged.

File: numpy/linalg/lapack_lite/make_lite.py [hunk 13]
```diff
@@ -266,13 +279,25 @@ def scrubF2CSource(c_file):
...
+    # Make sure that patch and f2c are found on path
+    ensure_executable('f2c')
+    ensure_executable('patch')
+
     wrapped_routines_file = sys.argv[1]
     lapack_src_dir = sys.argv[2]
-    output_dir = sys.argv[3]
+    output_dir = os.path.join(os.path.dirname(__file__), 'build')
+
+    shutil.rmtree(output_dir)
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/make_lite.py]: This needs a `try`/`except` to work on a fresh checkout

File: numpy/linalg/lapack_lite/make_lite.py [hunk 14]
```diff
@@ -232,6 +240,10 @@ def getWrappedRoutineNames(wrapped_routines_file):
 types = {'blas', 'lapack', 'd_lapack', 's_lapack', 'z_lapack', 'c_lapack', 'config'}
 
 def dumpRoutineNames(library, output_dir):
+    if not PY2:
+        os.makedirs(output_dir, exist_ok=True)
+    else:
+        os.makedirs(output_dir)
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/make_lite.py]: Why not move this to inside `main`, right after deleting the old one?

File: numpy/linalg/lapack_lite/clapack_scrub.py [hunk 15]
```diff
@@ -240,6 +277,8 @@ def scrubSource(source, nsteps=None, verbose=False):
              ('clean source', cleanSource),
              ('clean comments', cleanComments),
              ('replace dlamch_() calls', replaceDlamch),
+             ('remove definitions', removeSubroutinePrototypes),
+             ('remove builtin functions', removeBuiltinFunctions),
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/clapack_scrub.py]: Prototypes are declarations, not definitions, I think. Both messages should include the word "declaration"

File: numpy/linalg/lapack_lite/f2c.h (hunk 11 above)
Comment (eric-wieser) [reply numpy/linalg/lapack_lite/f2c.h]: Can you link to that random repo? Does this include the `Subroutine` declarations too?

File: numpy/linalg/lapack_lite/f2c_config.c (hunk 7 above)
Comment (eric-wieser) [reply numpy/linalg/lapack_lite/f2c_config.c]: Well, I rebuilt with my python 2 setup and this didn't change, so I have no idea what's going on, but I guess it's ok.

File: numpy/linalg/lapack_lite/make_lite.py [hunk 16]
```diff
@@ -266,13 +275,32 @@ def scrubF2CSource(c_file):
...
     lapack_src_dir = sys.argv[2]
-    output_dir = sys.argv[3]
+    output_dir = os.path.join(os.path.dirname(__file__), 'build')
+
+    try:
+        shutil.rmtree(output_dir)
+    except:
+        pass
+    try:
+        os.makedirs(output_dir)
```
Comment (eric-wieser) [inline numpy/linalg/lapack_lite/make_lite.py]: Would this ever fail? And if it does, is there any reason to catch it?

**Comment by eric-wieser** [comment]
Still would be good to replace "remove builtin functions" with "remove builtin function prototypes"

File: numpy/linalg/lapack_lite/f2c.h (hunk 11 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c.h]: I think it's [here](https://github.com/juanjosegarciaripoll/f2c/blob/master/include/f2c.h). But again, all that's required is to remove the inline prototypes and the compiler will check consistency. It's not required to trust that these definitions are correc.t

File: numpy/linalg/lapack_lite/f2c_config.c (hunk 7 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c_config.c]: Note: I did not compile f2c myself. I tried to compile a 64-bit version with anaconda, but there was a fault, so I gave up and downloaded the binary from netlib.

File: numpy/linalg/lapack_lite/f2c.h (hunk 11 above)
Comment (eric-wieser) [reply numpy/linalg/lapack_lite/f2c.h]: > the compiler will check consistency. I

Is that true if one is missing? (hence my question)

The one you link to [at juanjosegarciaripoll/f2c](https://github.com/juanjosegarciaripoll/f2c/blob/master/include/f2c.h) doesn't match the [debian package](
https://anonscm.debian.org/cgit/collab-maint/f2c.git/tree/f2c.h)...

**Comment by eric-wieser** [comment]
Found the source of that header file: https://anonscm.debian.org/cgit/collab-maint/libf2c2.git/tree/f2ch.add

Can you add a comment to the top of it saying that it comes from there (and/or replace it with that more authoritative source)

File: numpy/linalg/lapack_lite/make_lite.py (hunk 10 above)
Comment (eric-wieser) [reply numpy/linalg/lapack_lite/make_lite.py]: You succeeded in running this on windows? Nice!

File: numpy/linalg/lapack_lite/f2c.h (hunk 11 above)
Comment (xoviat) [reply numpy/linalg/lapack_lite/f2c.h]: > Is that true if one is missing? (hence my question)

Yes. There is only a problem when the prototypes are inconsistently defined across multiple files.

**Review by eric-wieser — APPROVED** [review]
Overall, looks very nice - much better than the patch I had in mind.

I'm still curious if you actually made this run on windows, but it looks good to merge either way.

I plan to squash these commits, unless you want to do that yourself

**Comment by charris** [comment]
If things are copied from Debian I worry about the license. The github version seems to retain the lucent license, which looks OK, but I'm not sure what happens with Debian. Maybe it is relicensed to the GNU license? And is taking a header file covered?

**Comment by eric-wieser** [comment]
The header file is also available directly from netlib, you just have to go through a zip to get it

**Comment by charris** [comment]
OK, great.

**Comment by eric-wieser** [comment]
@charris: Good to squash and merge then?

**Comment by charris** [comment]
Go for it.

**Comment by charris** [comment]
@eric-wieser I squashed and merged in order to do the backport, I want to release 1.14.1 this weekend.

@xoviat Thanks.
